pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTModuleProvidesDirective.java

```diff
@@ -0,0 +1,39 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import net.sourceforge.pmd.lang.ast.NodeStream;
+
+/**
+ * A "provides" directive of a {@linkplain ASTModuleDeclaration module declaration}.
+ *
+ * <pre class="grammar">
+ *
+ * ModuleProvidesDirective ::=
+ *     "provides" {@linkplain ASTClassOrInterfaceType ClassType}
+ *     "with" {@linkplain ASTClassOrInterfaceType ClassType} ( "," {@linkplain ASTClassOrInterfaceType ClassType} )*
+ *     ";"
+ *
+ * </pre>
+ */
+public final class ASTModuleProvidesDirective extends ASTModuleDirective {
+
+    ASTModuleProvidesDirective(int id) {
+        super(id, DirectiveType.PROVIDES);
+    }
+
+    @Override
+    protected <P, R> R acceptVisitor(JavaVisitor<? super P, ? extends R> visitor, P data) {
+        return visitor.visit(this, data);
+    }
+
+    public ASTClassOrInterfaceType getProvidedInterface() {
```

Comment:
Maybe call this "getService()"? - the JLS talks about "services" and "service providers"